aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp

```diff
@@ -0,0 +1,225 @@
+
+  if (input.dim() == 3 || input.size(0) == 1) {
+    AT_DISPATCH_QINT_TYPES(
+        input.scalar_type(), "quantized_adaptive_avg_pool2d", [&] {
+          auto input_data = input.data_ptr<scalar_t>();
+          auto output_data = output.data_ptr<scalar_t>();
+          adaptive_avg_pool2d_single_out_frame<scalar_t, underlying_t>(
+              input_data,
+              output_data,
+              sizeD,
+              isizeH,
+              isizeW,
+              osizeH,
+              osizeW,
+              istrideD,
+              istrideH,
+              istrideW);
+        });
+  } else {
+    int64_t istrideB = input.stride(-4);
+
+    AT_DISPATCH_QINT_TYPES(
+        input.scalar_type(), "quantized_adaptive_avg_pool2d", [&] {
+          auto input_data = input.data_ptr<scalar_t>();
+          auto output_data = output.data_ptr<scalar_t>();
+          adaptive_avg_pool2d_out_frame<scalar_t, underlying_t>(
+              input_data,
+              output_data,
+              sizeB,
+              sizeD,
+              isizeH,
+              isizeW,
+              osizeH,
+              osizeW,
+              istrideB,
+              istrideD,
+              istrideH,
+              istrideW);
+        });
+  }
+}
+
+std::vector<int64_t> get_output_shape(Tensor input, IntArrayRef output_size) {
+  for (int64_t i = 0; i < input.dim(); i++) {
+    TORCH_CHECK(
+        input.size(i) > 0,
+        "adaptive_avg_pooling2d(): expected input to have non-empty spatial "
+        "dimensions, but input has sizes ",
+        input.sizes(),
+        " with dimension ",
+        i,
+        " being empty");
+  }
+
+  TORCH_CHECK(
+      (input.dim() == 3 || input.dim() == 4),
+      "non-empty 3D or 4D (batch mode) tensor expected for input");
+
+  /* sizes */
+  int64_t sizeD = input.size(-3);
+  const auto osizeH = output_size[0];
+  const auto osizeW = output_size[1];
+
+  /* resize output */
+  std::vector<int64_t> output_shape;
+  int64_t sizeB = 0;
+  if (input.dim() == 3) {
+    output_shape = {sizeD, osizeH, osizeW};
+  } else {
+    sizeB = input.size(-4);
+    output_shape = {sizeB, sizeD, osizeH, osizeW};
+  }
+
+  return output_shape;
+}
+} // namespace
+
+Tensor& quantized_adaptive_avg_pool2d_out(
+    Tensor& output,
+    const Tensor& input,
+    IntArrayRef output_size) {
+  const auto output_shape = get_output_shape(input, output_size);
+  TORCH_CHECK(
+      output.is_quantized() && output.sizes() == output_shape,
+      "Output Tensor must be quantized and have a shape of ",
+      "{ ",
+      output_shape,
+      " }.");
+  adaptive_avg_pool2d_out_template(output, input, output_shape);
+  return output;
+}
+
+Tensor quantized_adaptive_avg_pool2d(
+    const at::Tensor& input,
+    IntArrayRef output_size) {
+  const auto output_shape = get_output_shape(input, output_size);
+  Tensor output = at::_empty_affine_quantized(
+      output_shape, input.options(), input.q_scale(), input.q_zero_point());
```

Comment:
This does not preserve the layout of the input tensor, meaning you will see poor performance in downstream operators